File: frontend/src/types/finishedProduct.ts
import type {
  FinalBill,
} from "./finalBill";

import type {
  ProductionMaterial,
  ProductionOperation,
  ProductionOrder,
} from "./production";


export interface FinishedProduct {
  id: number;

  finished_product_number: string;

  product_name: string;

  unit: string;

  product_master_id: number | null;

  production_order_id: number;

  finished_goods_receipt_id: number;

  created_by: number;

  created_at: string;
}


export interface FinishedProductMaster {
  id: number;

  product_code: string;

  product_name: string;

  description: string | null;

  category: string;

  unit: string;

  hsn_code: string;

  gst_percentage: string;
}


export interface FinishedProductCustomer {
  id: number;

  customer_code: string;

  company_name: string;

  contact_person: string | null;

  email: string | null;

  phone: string | null;

  gst_number: string | null;

  address: string | null;

  city: string | null;

  state: string | null;

  pincode: string | null;
}


export interface FinishedProductEnquiry {
  id: number;

  enquiry_number: string;

  enquiry_date: string;

  customer_id: number;

  company_name: string;

  contact_person: string | null;

  phone: string | null;

  email: string | null;

  gst_number: string | null;

  address: string | null;

  city: string | null;

  state: string | null;

  pincode: string | null;

  machine_name: string | null;

  machine_model: string | null;

  application: string | null;

  quantity: number | null;

  requirements: string | null;

  remarks: string | null;

  status: string;
}


export interface FinishedProductProforma {
  id: number;

  proforma_number: string;

  proforma_date: string;

  enquiry_id: number;

  customer_id: number;

  company_name: string;

  subtotal: string;

  discount_amount: string;

  taxable_amount: string;

  tax_amount: string;

  grand_total: string;

  status: string;
}


export interface FinishedProductShopFloorIssue {
  id: number;

  issue_number: string;

  production_order_id: number;

  production_material_id: number;

  product_id: number;

  quantity_issued: string;

  unit_cost: string;

  total_cost: string;

  stock_before: string;

  stock_after: string;

  issued_by: number;

  issued_at: string;

  remarks: string | null;
}


export interface FinishedProductGoodsReceipt {
  id: number;

  receipt_number: string;

  production_order_id: number;

  product_id: number | null;

  quantity_received: string;

  stock_before: string;

  stock_after: string;

  received_by: number;

  received_at: string;

  remarks: string | null;
}


export interface FinishedProductCostSummary {
  actual_material_cost: string;

  actual_operation_cost: string;

  actual_production_cost: string;

  finished_quantity: string;

  cost_per_unit: string;
}


export interface FinishedProductBilling {
  original_invoice: FinalBill | null;

  effective_invoice: FinalBill | null;

  revisions: FinalBill[];

  credit_notes: FinalBill[];
}


export interface FinishedProductTraceability {
  finished_product: FinishedProduct;

  /*
   * Legacy only.
   * New custom-manufactured products normally have null here.
   */
  product_master: FinishedProductMaster | null;

  customer: FinishedProductCustomer;

  enquiry: FinishedProductEnquiry;

  proforma: FinishedProductProforma;

  production_order: ProductionOrder;

  production_materials: ProductionMaterial[];

  shop_floor_issues:
    FinishedProductShopFloorIssue[];

  production_operations:
    ProductionOperation[];

  finished_goods_receipt:
    FinishedProductGoodsReceipt;

  cost_summary:
    FinishedProductCostSummary;

  billing:
    FinishedProductBilling;
}

File: backend/app/schemas/finished_product.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)

from app.schemas.final_bill import (
    FinalBillResponse,
)
from app.schemas.finished_goods_receipt import (
    FinishedGoodsReceiptResponse,
)
from app.schemas.production import (
    ProductionMaterialResponse,
    ProductionOperationResponse,
)
from app.schemas.shop_floor_issue import (
    ShopFloorIssueResponse,
)


# ============================================================
# BASIC FINISHED PRODUCT
# ============================================================


class FinishedProductResponse(BaseModel):
    id: int

    finished_product_number: str

    # Manufactured output source of truth.
    product_name: str

    unit: str

    # Legacy purchased Product relationship.
    product_master_id: int | None = None

    production_order_id: int

    finished_goods_receipt_id: int

    created_by: int

    created_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# LEGACY PRODUCT MASTER
# ============================================================


class FinishedProductMasterResponse(BaseModel):
    id: int

    product_code: str

    product_name: str

    description: str | None = None

    category: str

    unit: str

    hsn_code: str

    gst_percentage: Decimal

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# CUSTOMER
# ============================================================


class FinishedProductCustomerResponse(BaseModel):
    id: int

    customer_code: str

    company_name: str

    contact_person: str | None = None

    email: str | None = None

    phone: str | None = None

    gst_number: str | None = None

    address: str | None = None

    city: str | None = None

    state: str | None = None

    pincode: str | None = None

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# ENQUIRY
# ============================================================


class FinishedProductEnquiryResponse(BaseModel):
    id: int

    enquiry_number: str

    enquiry_date: date

    customer_id: int

    company_name: str

    contact_person: str | None = None

    phone: str | None = None

    email: str | None = None

    gst_number: str | None = None

    address: str | None = None

    city: str | None = None

    state: str | None = None

    pincode: str | None = None

    machine_name: str | None = None

    machine_model: str | None = None

    application: str | None = None

    quantity: int | None = None

    requirements: str | None = None

    remarks: str | None = None

    status: str

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# PROFORMA
# ============================================================


class FinishedProductProformaResponse(BaseModel):
    id: int

    proforma_number: str

    proforma_date: date

    enquiry_id: int

    customer_id: int

    company_name: str

    subtotal: Decimal

    discount_amount: Decimal

    taxable_amount: Decimal

    tax_amount: Decimal

    grand_total: Decimal

    status: str

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# PRODUCTION ORDER
# ============================================================


class FinishedProductProductionResponse(BaseModel):
    id: int

    production_number: str

    proforma_id: int

    proforma_item_id: int | None = None

    product_name: str

    unit: str

    # Legacy purchased Product relationship.
    product_id: int | None = None

    quantity: int

    status: str

    planned_start_date: date | None = None

    actual_start_date: date | None = None

    actual_end_date: date | None = None

    notes: str | None = None

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


# ============================================================
# COST SUMMARY
# ============================================================


class FinishedProductCostSummaryResponse(BaseModel):
    actual_material_cost: Decimal = Field(
        default=Decimal("0.00"),
    )

    actual_operation_cost: Decimal = Field(
        default=Decimal("0.00"),
    )

    actual_production_cost: Decimal = Field(
        default=Decimal("0.00"),
    )

    finished_quantity: Decimal = Field(
        default=Decimal("0.00"),
    )

    cost_per_unit: Decimal = Field(
        default=Decimal("0.00"),
    )


# ============================================================
# BILLING TRACEABILITY
# ============================================================


class FinishedProductBillingResponse(BaseModel):
    original_invoice: FinalBillResponse | None = None

    effective_invoice: FinalBillResponse | None = None

    revisions: list[FinalBillResponse] = Field(
        default_factory=list,
    )

    credit_notes: list[FinalBillResponse] = Field(
        default_factory=list,
    )


# ============================================================
# COMPLETE TRACEABILITY
# ============================================================


class FinishedProductTraceabilityResponse(BaseModel):
    finished_product: FinishedProductResponse

    # Legacy finished products may have Product Master data.
    # New manufactured products normally do not.
    product_master: FinishedProductMasterResponse | None = None

    customer: FinishedProductCustomerResponse

    enquiry: FinishedProductEnquiryResponse

    proforma: FinishedProductProformaResponse

    production_order: FinishedProductProductionResponse

    production_materials: list[ProductionMaterialResponse] = Field(
        default_factory=list,
    )

    shop_floor_issues: list[ShopFloorIssueResponse] = Field(
        default_factory=list,
    )

    production_operations: list[ProductionOperationResponse] = Field(
        default_factory=list,
    )

    finished_goods_receipt: FinishedGoodsReceiptResponse

    cost_summary: FinishedProductCostSummaryResponse

    billing: FinishedProductBillingResponse = Field(
        default_factory=FinishedProductBillingResponse,
    )